django/test/testcases.py

```diff
@@ -150,6 +150,48 @@ def __call__(self):
         )
 
 
+class _AssertSignalSentContext(object):
+    def __init__(self, test_case, signal, required_kwargs=None):
+        self.test_case = test_case
+        self.signal = signal
+        self.required_kwargs = required_kwargs
+
+    def _listener(self, sender, **kwargs):
+        self.signal_sent = True
+        self.received_kwargs = kwargs
+        self.sender = sender
+
+    def __enter__(self):
+        self.signal_sent = False
+        self.received_kwargs = {}
+        self.sender = None
+        self.signal.connect(self._listener)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.signal.disconnect(self._listener)
+        if not self.signal_sent:
+            self.test_case.fail('Signal was not sent.')
```

Comment:
How do you feel about this `# HACK` :-D:
```diff
diff --git a/django/dispatch/dispatcher.py b/django/dispatch/dispatcher.py
index 30f339a..ef0e266 100644
--- a/django/dispatch/dispatcher.py
+++ b/django/dispatch/dispatcher.py
@@ -53,6 +53,8 @@ class Signal(object):
         # .disconnect() is called and populated on send().
         self.sender_receivers_cache = weakref.WeakKeyDictionary() if use_caching else {}
         self._dead_receivers = False
+        import traceback
+        self.name = traceback.extract_stack()[-2].line.split('=')[0].strip()
 
     def connect(self, receiver, sender=None, weak=True, dispatch_uid=None):
         """
```
It looks like it would work for the built-in signals at least.
(from http://stackoverflow.com/questions/14413025/how-can-i-find-out-where-an-object-has-been-instantiated)